framework/editor/InputTarget.kt

```diff
@@ -248,26 +248,26 @@ internal class InputTarget constructor(
     internal fun clearStatus() = Service.status.clear(TEXT_CURSOR_POSITION)
 
     private fun mayScrollToCursor() {
-        fun mayScrollToCoordinate(x: Int, y: Int, padding: Int = 0) {
-            val left = textAreaBounds.left.toInt() + padding
-            val right = textAreaBounds.right.toInt() - padding
-            val top = textAreaBounds.top.toInt() + padding
-            val bottom = textAreaBounds.bottom.toInt() - padding
+        fun mayScrollToCoordinate(x: Int, y: Int, padding: Dp = 0.dp) {
+            val left = textAreaBounds.left.toInt()
+            val right = textAreaBounds.right.toInt()
+            val top = textAreaBounds.top.toInt()
+            val bottom = textAreaBounds.bottom.toInt()
             if (x < left) coroutines.launch {
                 horScroller.scrollTo(horScroller.value + ((x - left) * density).toInt())
             } else if (x > right) coroutines.launch {
                 horScroller.scrollTo(horScroller.value + ((x - right) * density).toInt())
             }
-            if (y < top) verScroller.updateOffsetBy((y - top).dp)
-            else if (y > bottom) verScroller.updateOffsetBy((y - bottom).dp)
+            if (y <= top) verScroller.updateOffsetBy((y - top).dp - padding)
```

Comment:
If the cursor is at the top or bottom exactly (not just further up/down than the top/bottom), we should scroll to that line. updateOffsetBy already coerces this to the maximum offset, so we can only go as far down as the maximum depth.